src/callbacks_stage/positivity_shallow_water.jl

```diff
@@ -0,0 +1,82 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+ # Since these FMAs can increase the performance of many numerical algorithms,
+ # we need to opt-in explicitly.
+ # See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+ @muladd begin
+
+
+  """
+      PositivityPreservingLimiterShallowWater(; threshold, variables)
+ 
+  This fully-discrete positivity-preserving limiter is based on the work of 
+  - Zhang, Shu (2011)
+    Maximum-principle-satisfying and positivity-preserving high-order schemes
+    for conservation laws: survey and new developments
+    [doi: 10.1098/rspa.2011.0153](https://doi.org/10.1098/rspa.2011.0153)
+ 
+  It is specifically designed for `ShallowWaterEquations`.
+  The limiter is applied to all scalar `variables` in their given order
+  using the associated `thresholds` to determine the minimal acceptable values.
+  The order of the `variables` is important and might have a strong influence
+  on the robustness. 
+  As opposed to the default version of the limiter, nodes with a waterheight below the threshold
+  are treated different. To avoid numerical problems caused by velocities close to zero in that case,
+  the velocity is cut off, as well, as the node can be identified as dry. The special feature of the
+  `ShallowWaterEquations`` used here is that the bottom topography is stored in `u`. However, this value
+  should not be changed, much less limited. That is why, the `set_node_vars!` function is not used
+  for the last conservation variable.
+  Afterwards for safety reasons, the wet/dry threshold is applied again on all the DG nodes of freedom
+  in order to avoid dry nodes. In the case where value_mean < threshold before applying
+  limiter, there could still be dry nodes afterwards due to the logic of the limiter.
```

Comment:
```suggestion
  As opposed to the standard version of the [PositivityPreservingLimiterZhangShu](@ref), 
  nodes with a water height below the `threshold_limiter` are treated in a special way. 
  To avoid numerical problems caused by velocities close to zero,
  the velocity is cut off, such that the node can be identified as "dry". The special feature of the
  `ShallowWaterEquations` used here is that the bottom topography is stored as an additional
  quantity in the solution vector `u`. However, the value of the bottom topography
  should not be changed, much less limited. That is why, the `set_node_vars!` function is not applied
  to the last conservation variable.
  After the limiting process is applied to all degrees of freedom, for safety reasons,
  the wet/dry threshold is applied again on all the DG nodes in order to avoid dry nodes. 
  In the case where value_mean < threshold before applying
  limiter, there could still be dry nodes afterwards due to the logic of the limiter.
```